src/auth/auth_gitlab.py

```diff
@@ -0,0 +1,339 @@
+                    del self.user_states[userData]
+
+    def dump_sessions_to_file(self):
+        if self.gitlab_dump:
+            dumpFile = open(self.gitlab_dump, "w")
+            dumpFile.write(escape.json_encode(self.user_states))
+            dumpFile.close()
+            LOGGER.debug("Dumped state to file %s" % self.gitlab_dump)
+
+    def do_update_user(self, user, access_token):
+        self.user_states[user]['updating'] = True
+        tornado.ioloop.IOLoop.current().spawn_callback(self.update_user_state, user, access_token)
+
+    def do_update_groups(self, user, access_token):
+        self.user_states[user]['updating'] = True
+        tornado.ioloop.IOLoop.current().spawn_callback(self.update_group_list, user, access_token)
+
+    @gen.coroutine
+    def update_group_list(self, user, access_token):
+        group_list = yield self.read_groups(access_token)
+        if group_list is None:
+            LOGGER.error("Failed to refresh groups for %s" % user)
+            self.user_states[user]['state'] = "error"
+        else:
+            LOGGER.info("Groups for %s refreshed: " % user + str(group_list))
+            self.user_states[user]['groups'] = group_list
+        now = time.time()
+        self.user_states[user]['updating'] = False
+        self.user_states[user]['updated'] = now
+        self.user_states[user]['visit'] = now
+        self.clean_expired_sessions()
+        self.dump_sessions_to_file()
+        return
+
+    @gen.coroutine
+    def update_user_state(self, user, access_token):
+        user_state = yield self.fetch_user(access_token)
+        if user_state is None:
+            LOGGER.error("Failed to fetch user %s" % user)
+            self.user_states[user]['state'] = "error"
+        else:
+            LOGGER.info("User %s refreshed: " % user + str(user_state))
+            self.user_states[user] = {**self.user_states[user], **user_state}
+        now = time.time()
+        self.user_states[user]['updating'] = False
+        self.user_states[user]['updated'] = now
+        self.user_states[user]['visit'] = now
+        self.clean_expired_sessions()
+        self.dump_sessions_to_file()
+        return
+
+    @gen.coroutine
+    def read_groups(self, access_token):
+        args = {
+            'access_token': access_token,
+            'all_available': 'false',
+            'per_page': 100,
+        }
+        if not self.gitlab_group_search is None:
+            args['search'] = self.gitlab_group_search
+
+        group_list_future = self.oauth2_request(
+            self._OAUTH_GITLAB_GROUPS % self._GITLAB_PREFIX,
+            **args
+        )
+
+        group_list = yield group_list_future
+
+        if group_list is None:
+            return None
+
+        groups = []
+        for group in group_list:
+            if group.get('full_path'):
+                groups.append(group['full_path'])
+
+        return groups
+
+    @gen.coroutine
+    def validate_user(self, code, request_handler):
```

Comment:
I would suggest to call it fetch/load user (or anything similar). As far as I can understand from the code. the main purpose of the function is loading the data, not data validation

Reply:
agree